CMakeLists.txt

```diff
@@ -130,7 +117,16 @@ IF (CMAKE_C_COMPILER_ID MATCHES "^Clang$")
   SET(CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG} -Wshadow")
   SET(CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG} -Wmissing-prototypes")
   SET(CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG} -Wcast-qual")
-ENDIF (CMAKE_C_COMPILER_ID MATCHES "^Clang$")
+  # Place the functions and data into separate sections, allowing the linker
+  # to garbage collect the unused ones.
+  SET(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -ffunction-sections -fdata-sections")
+  SET(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -Wl,--gc-sections")
+  SET(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -Wl,--gc-sections")
+  # In Debug builds print the actual sections which are discarded
+  SET(CMAKE_EXE_LINKER_FLAGS_DEBUG "${CMAKE_EXE_LINKER_FLAGS_DEBUG} -Wl,--print-gc-sections")
+  SET(CMAKE_SHARED_LINKER_FLAGS_DEBUG "${CMAKE_SHARED_LINKER_FLAGS_DEBUG} -Wl,--print-gc-sections")
```

Comment:
I doubt anyone who is not actually debugging build problems with libarchive cares about this.  I think I might leave these lines here but commented out so that people who need to debug build problems can get the extra information.

Reply:
What's why I've enabled it only in debug builds. Is that too much?

Reply:
:shrug: I think it might be too much.  Folks who are including a bunch of libraries via CMake `add_subdirectory` and building the entire resulting system in debug mode are probably not interested in a list of things that one of those libraries is _not_ linking.